Introduce log levels for DiagnosticLogs

sahandilshan opened the pull request:
## Purpose
* $subject
* This was introduced to show two different log levels for the diagnostic logs

sahandilshan committed:
Introduce log levels for DiagnosticLogs

chamathns commented on core/org.wso2.carbon.utils/src/main/java/org/wso2/carbon/utils/DiagnosticLog.java:
```diff
@@ -102,4 +121,14 @@ public Map<String, Object> getConfigurations() {
+
+    public enum LogLevel {
+        BASIC, // This level is intended for App Developers (and Internal Developers).
+        ADVANCE // This level is intended only for Internal Developers.
```
shall we use `ADVANCED`?

sahandilshan replied:
Fixed with cbe430e57c49d1390ee6c9d498c1b26cfaab709a

chamathns commented on core/org.wso2.carbon.utils/src/main/java/org/wso2/carbon/utils/DiagnosticLog.java:
```diff
@@ -102,4 +121,14 @@ public Map<String, Object> getConfigurations() {
+
+        return logLevel;
+    }
+
```
add javadoc comments

sahandilshan replied:
Fixed with 5a88f68259a09bda7d564506434c02818d7e6c74

chamathns commented on core/org.wso2.carbon.utils/src/main/java/org/wso2/carbon/utils/DiagnosticLog.java:
```diff
@@ -102,4 +121,14 @@ public Map<String, Object> getConfigurations() {
+    }
+
+    public enum LogLevel {
+        BASIC, // This level is intended for App Developers (and Internal Developers).
```
Leave a blank line

sahandilshan replied:
Fixed with cbe430e57c49d1390ee6c9d498c1b26cfaab709a

chamathns commented on core/org.wso2.carbon.utils/src/main/java/org/wso2/carbon/utils/DiagnosticLog.java:
```diff
@@ -102,4 +121,14 @@ public Map<String, Object> getConfigurations() {
         return configurations;
     }
+
+    public LogLevel getPersona() {
```
shall we use `getLogLevel`?

sahandilshan replied:
Fixed with cbe430e57c49d1390ee6c9d498c1b26cfaab709a

sahandilshan committed:
Fix formatting issues

pulasthi7 commented on core/org.wso2.carbon.utils/src/main/java/org/wso2/carbon/utils/DiagnosticLog.java:
```diff
@@ -51,6 +53,24 @@ public DiagnosticLog(String logId, Instant recordedAt, String requestId, String
+
+    public DiagnosticLog(String logId, Instant recordedAt, String requestId, String flowId,
+            String resultStatus, String resultMessage, String actionId, String componentId,
+            Map<String, Object> input, Map<String, Object> configurations, LogLevel logLevel) {
```
are all these params mandatory? if not please consider using builder pattern

sahandilshan replied:
Fixed with 3ea6c8e08f8f772ddd97b61d80efa645f86728b9

sahandilshan committed:
Add the builder class to create the object

sahandilshan committed:
Initialise null values before creating object

chamathns commented on core/org.wso2.carbon.utils/src/main/java/org/wso2/carbon/utils/DiagnosticLog.java:
```diff
@@ -51,6 +61,24 @@ public DiagnosticLog(String logId, Instant recordedAt, String requestId, String
+        this.input = input;
+        this.configurations = configurations;
+        this.logLevel = logLevel;
     }
```
shall we change the new public constructor to use a static nested builder class?

eg: public DiagnosticLog(PublicDiagnosticLogEventBuilder builder)

sahandilshan replied:
Fixed with a3cd4390c2d8ceeaeb711f9a6580e000ceb7c4b7 and also make the constructor private as well

sahandilshan committed:
Make the DiagnosticLog constructor private

Make the DiagnosticLog constructor private so that to create the object we have to use the builder

chamathns commented on core/org.wso2.carbon.utils/src/main/java/org/wso2/carbon/utils/DiagnosticLog.java:
```diff
@@ -102,4 +128,208 @@ public Map<String, Object> getConfigurations() {
+         */
+        public DiagnosticLog build() {
+
+            if (this.logId != null) {
```
not null? shouldn't we add a UUID if it's null?

sahandilshan replied:
Fixed with b4971ec182001f59d683ea61b29217e4089ab118

AnuradhaSK commented on core/org.wso2.carbon.utils/src/main/java/org/wso2/carbon/utils/DiagnosticLogUtils.java:
```diff
@@ -0,0 +1,51 @@
+package org.wso2.carbon.utils;
```
missing license header

sahandilshan replied:
This has been fixed now (Removed the Util class).

AnuradhaSK commented on core/org.wso2.carbon.utils/src/main/java/org/wso2/carbon/utils/DiagnosticLogUtils.java:
```diff
@@ -0,0 +1,51 @@
+ */
+public class DiagnosticLogUtils {
+
+    private static final Log log = LogFactory.getLog(DiagnosticLogUtils.class);
```
```suggestion
    private static final Log LOG = LogFactory.getLog(DiagnosticLogUtils.class);
```
use all capital for final variables

pulasthi7 commented on core/org.wso2.carbon.utils/src/main/java/org/wso2/carbon/utils/DiagnosticLog.java:
```diff
@@ -102,4 +128,208 @@ public Map<String, Object> getConfigurations() {
+                logId = UUID.randomUUID().toString();
+            }
+            if (this.recordedAt == null) {
+                recordedAt = parseDateTime(Instant.now().toString());
```
Instant -> String -> Instant conversion. seems not necessary and can directly use Instant.now()

sahandilshan replied:
This has been fixed now

sahandilshan committed:
Modify the builder class with Mandatory params

chamathns commented on core/org.wso2.carbon.utils/src/main/java/org/wso2/carbon/utils/DiagnosticLog.java:
```diff
@@ -102,4 +127,246 @@ public Map<String, Object> getConfigurations() {
+        SUCCESS, // Successful execution.
+        FAILED, // Failed execution.
+        PENDING, // Pending execution.
+        NONE // No result status.
```
do we need these statuses? I feel like this could lead to inconsistencies with the usages. Currently, we add "SUCCESS" status to all the non-error flows. If we introduce something like PENDING, NONE, we would need to change the already existing flows as well. But, I don't really see a need to include these. WDYT?

cc: @pulasthi7

sahandilshan replied:
removed unwanted status with 6d2d59fb69b1ebac978aa76cacc1d441b6c0d204

sahandilshan committed:
Remove unwanted status from the ResultStatus

sahandilshan committed:
Allow to add all types of objects to the map
